app/grandchallenge/components/views.py

```diff
@@ -288,3 +286,77 @@ def get_success_url(self):
 
     def get_success_message(self, cleaned_data):
         return f"{self.object._meta.verbose_name.title()} was successfully deleted"
+
+
+class BaseModelOptions(TextChoices):
+    ARCHIVE = Archive._meta.model_name
+    READER_STUDY = ReaderStudy._meta.model_name
+
+
+class CivSetListView(
+    LoginRequiredMixin, PermissionListMixin, PaginatedTableListView
+):
+    model = None
+    permission_required = None
+    raise_exception = True
+    template_name = "components/civ_set_list.html"
+    row_template = "components/civ_set_row.html"
+    search_fields = [
+        "pk",
+        "values__interface__title",
+        "values__image__name",
+        "values__file",
+    ]
+    extra_columns = []
+    text_align = "left"
+    default_sort_order = "asc"
+    base_object_lookup = None
+    extra_prefetch_fields = []
+    columns = [
+        Column(title="Values"),
+        Column(title="View"),
+        Column(title="Edit"),
+        Column(title="Remove"),
+    ]
+
+    def __init__(self):
+        if self.extra_columns:
+            self.columns = [*self.extra_columns, *self.columns]

```

Comment:
This is mixing class and instance properties. Rather than adding `extra_prefetch_fields` you could remove that altogether and I think do something like:

```python
class ReaderStudyDisplaySetList(CivSetListView):
   columns = [
         *CivSetListView.columns,
         Column(title="DisplaySet ID", sort_field="pk"),
         Column(title="Order", sort_field="order"),
     ]
```

That way, `columns` remains a Class property of `ReaderStudyDisplaySetList` rather than an instance property - there may be some assumptions of that (we've run into bugs before due to confusion there).

Reply:
Ah ok, that makes sense for columns. For get_queryset with the prefetch fields, I'm not sure yet what's best. To avoid calling instance properties in the base view, I think it would actually be best to move the entire `get_queryset` method to the instance and set a property `prefetch_fields` on the base view to call from the instance method. Would you agree?

If we leave the queryset method on the base view (doing the select_related and the prefetching of the common fields in the base view and then add an extra prefetch_related in the instance method), we'd add an extra db call, which would be less efficient db-query wise than doing it all in one go from what I understand. So seems to me that leaving `get_queryset` to the instances is better, even if that means a bit of code duplication.

Reply:
This is about mixing of class and instance properties - the way you have it now means that columns is sometimes a class attribute, sometimes an instance attribute. It means the class attribute is no longer safe to use. So, it means either revert to having columns as a class property, or to turn everything into an instance property. The former is less work.

I think the new way is clearer as there is less indirection, but there is no need to call `CivSetListView` in an instance method, you can use `super()` here: `fields_to_prefetch = super().prefetch_fields`.

One thing you could consider is doing:

```python
class CivSetListView:
  def get_queryset(self):
    queryset = super().get_queryset()
    return queryset.prefetch_related("values", "values__image", "values__interface")

class ReaderStudy:
  def get_queryset(self):
    queryset = super().get_queryset()
    return queryset.prefetch_related("answers").filter(reader_study=self.base_object).select_related("reader_study")
```

I think that works but I'm unsure.

Reply:
> One thing you could consider is doing:

Hm, I'm not sure I like this better than what I have now. I feel like it's clearer if there is only one place of `get_queryset` to look at rather than two. Also, I think this way would be slightly less efficient, right? We'd be doing an unnecessary extra DB query, plus we're calling select_related after the prefetch_related call...

Reply:
> So, it means either revert to having columns as a class property, or to turn everything into an instance property.

Now I'm a bit confused. I thought that in my current implementation `columns` _is_ a class rather than an instance property of the list views. They're defined directly in the class definition, not in the init method any longer. Am I misunderstanding your comment?

Reply:
> Hm, I'm not sure I like this better than what I have now. I feel like it's clearer if there is only one place of get_queryset to look at rather than two. 

This is more like how inheritance should work. With it as it is now the consumers of the View need to know about the custom internal expectations (they have to call `prefetch_related` with an attribute). With what I propose, the consumers don't care, and are only concerned with their own concrete implementation of the view. This allows for looser coupling.

> We'd be doing an unnecessary extra DB query, plus we're calling select_related after the prefetch_related call

AFAIK that shouldn't matter as selects and prefetches get bundled. But maybe they don't? You can check with the `.query` attribute of the queryset.


> Now I'm a bit confused.

Sorry, I wrote the comment before realising that you had made changes. That part looks good now.

Reply:
> AFAIK that shouldn't matter as selects and prefetches get bundled. But maybe they don't? You can check with the `.query` attribute of the queryset.

Looks like you are right. The generated queries are identical. I will update the code to use what you proposed.